ENH,BLD: Dynamically generate a .pyi file with platform-specific type hints

BvB93 opened the pull request:
An issue which has plagued the typing numpy for some time are objects whose value or, even worse, 
whose type is platform-specific. The most notorious example here is `np.int_`, the latter being an 
alias for either `np.int32` (on windows for example) or `np.int64`.

It is desirable to have some sort of mechanism for annotating these platform-specifics; in order to 
accomplish this the PR does two things:
* It adds a script which dynamically generates a `.pyi` file. This script is run at the very end of 
  `setup_package()`, ensuring that a fresh `.pyi` file is generated after every numpy installation 
  (similar to `np.version`).
* It adds a (second) new submodule to `np.typing` with valid character codes for each `np.generic` 
  subclass,  the latter also containing a number of platform specifics in the case of `np.number` 
  subclasses.

Pinging @person142 as this would help with a few issues encountered in https://github.com/numpy/numpy/pull/16759.

EDIT: See https://github.com/numpy/numpy/pull/17843 for an alternative approach using a mypy plugin.

Commit:
ENH,BLD: Automatically generate system-specific aliases upon installing numpy

Commit:
MAINT: Add the dynamically generated .pyi file to .gitignore

Commit:
TST: Add a test for dynamically generated type aliases

Commit:
MAINT: Renamed `_number_aliases` to `_dynamic_types`

Commit:
MAINT: Replace `Union[int32, int64]` with `int_

rgommers commented:
Thanks @BvB93, this sounds like a good idea. 

> the latter also containing a number of platform specifics in the case of `np.number` subclasses.

I'm actually failing to spot the platform-specific bit. E.g. a common type of dynamic check like `is_32_bit = np.intp(0).itemsize < 8`. What am I overlooking?

BvB93 commented:
> I'm actually failing to spot the platform-specific bit. E.g. a common type of dynamic check `like is_32_bit = np.intp(0).itemsize < 8`. What am I overlooking?

They keyword here is "dynamic". The example you provided works just during runtime, but for static 
type checking it has to be, well, defined statically. The latter is somewhat tricky if its type (or value) 
is platform specific, hence this PR.

As a side note, the CI failures suggest that there are still some problems with the current implementation; 
I'll take a closer look at it tomorrow.

Commit:
TST: Move and update the tests

Commit:
MAINT: Added the `clongdouble` type alias

Commit:
MAINT: Fixed an issue where the dynamic .pyi file wasn't properly generated

eric-wieser commented on numpy/typing/_dynamic_types.py:
```diff
@@ -0,0 +1,51 @@
+"""
```
Can we call this `_type_aliases.py` or `_platform_aliases.py` instead?

BvB93 replied:
Renamed as of https://github.com/numpy/numpy/pull/17514/commits/59d730e5f674e75859e2dc322802d9f126a32173.

eric-wieser commented on tools/generate_type_hints.py:
```diff
@@ -0,0 +1,221 @@
+"""
+
+
+def _file_to_context(
+    file: Union[IO[str], _AnyPath], *args: Any, **kwargs: Any
+) -> ContextManager[IO[str]]:
+    """Create a context manager from a path- or file-like object."""
+    try:
+        return open(file, *args, **kwargs)
+    except TypeError:
+        return nullcontext(file)
+
+
+def generate_alias(file: Union[IO[str], _AnyPath] = sys.stdout) -> None:
```
I'd be inclined to remove this and either:

* Just use `open` directly in `generate_aliases`, since the only two places you call it pass strings
* Use `open` in the two callers, to match python APIs like `pickle.dump` and `json.dump`, both of which expect already open file handles.

BvB93 replied:
Sure, why not. I went with the second option in https://github.com/numpy/numpy/pull/17514/commits/368fc435bd68c1d098201bf3947efbd25a67babc.

eric-wieser commented on tools/generate_type_hints.py:
```diff
@@ -0,0 +1,221 @@
+            # Plan B in case an untyped class is encountered
+            type_alias[name] = "Any"
+
+    # Generate the docstring
+    now = dt.datetime.now()
+    docstring = f'"""THIS FILE WAS AUTOMATICALLY GENERATED ON {now}."""'
```
I think we've had complaints in the past about unreproducible builds caused by putting timestamps in things, so I'd lean towards not doing so here.

BvB93 replied:
Done as of https://github.com/numpy/numpy/pull/17514/commits/5887e625eac696d03299bcc926cd1b5c3d7d5a3a.

Commit:
MAINT,TST: Fixed an issue where the dynamic .pyi file wasn't generated by runtest.py

Commit:
REV: Remove the date and time from the docstring

Commit:
MAINT: Ensure that `generate_alias` can only accept file-like objects

Commit:
MAINT: Renamed `dynamic_types` to `platform_aliases`

Commit:
MAINT: Renamed `dynamic_types` to `platform_aliases`

BvB93 commented:
Converting this into a draft for now, as there are still some issues getting the tests running properly.

Commit:
BLD: Write absolute path of `_platform_aliases.pyi` to the build logfile

eric-wieser reviewed (commented):
I think it would be worth a detailed explanation somewhat about exactly what is going on with the aliases:

* `char` through `longlong` are the native types used inside numpy corresponding to C types
* `int8` through `int64` are both:
  * aliases for the above types
  * the default __name__ of all but one of these types (`intc` or `longlong` depending on platform)
* the typing stubs are written as if the sized types are the native ones
* these files generate aliases in the reverse direction
* the reverse aliases are not quite accurate - they suggest that two of `intc, int_, longlong` are identical, when in fact there are distinct scalar types

BvB93 commented:
> I think it would be worth a detailed explanation somewhat about exactly what is going on with the aliases:

Good point, I'll update the PR description in a bit.

> the typing stubs are written as if the sized types are the native ones

To be fair, this is consist with how the `number` subclasses are treated during runtime 
(at least on the python side), where their C names are treated as aliases for their corresponding 
sized type rather than the other way around.

```python
In [2]: np.byte
Out[2]: numpy.int8
```

> the reverse aliases are not quite accurate - they suggest that two of `intc`, `int_`, `longlong` are identical, when in fact there are distinct scalar types

I'm not quite sure how you arrived at this conclusion, as can be seen below the only two integer types 
which are considered identical by definition are `intp` and `int0`.

https://github.com/numpy/numpy/blob/e6a5514cae950f9336866ac39b4880f8b80bcce0/tools/generate_type_hints.py#L138-L143

Running the script locally yields the following output (MacOS):

``` python
byte = int8
short = int16
intc = int32
intp = int0 = int64
int_ = int64
longlong = Any
```

eric-wieser commented:
> Running the script locally yields the following output (MacOS):
> 
> ```python
> byte = int8
> short = int16
> intc = int32
> intp = int0 = int64
> int_ = int64
> longlong = Any
> ```

Ah, I wasn't aware that `longlong = Any` was the result of this script.

BvB93 commented:
> Ah, I wasn't aware that `longlong = Any` was the result of this script.

Yes, `Any` is used as a fallback value here in case the aliased typed is still unannotated (or if it isn't an alias at all!).

Commit:
MAINT: print the path the `_platform_aliases.pyi` (debugging)

BvB93 commented:
So I'm having some trouble to get this working with the likes of `pip install -V .` (as is done in the tests).

The part I'm struggling with is accessing the content of `np.core` ("Importing the numpy C-extensions failed") 
in order to resolve the aliases of all the `np.generic` subclasses. I suspect that I'll have to hook into `np.distutils`
 _somewhere_ to get this working, but I'm not familiar enough with the whole setup process to narrow it down any further.

Could anyone provide some more insight here?

mattip commented on runtests.py:
```diff
@@ -165,6 +165,7 @@ def main(argv):
         sys.path.insert(0, site_dir)
         sys.path.insert(0, site_dir_noarch)
         os.environ['PYTHONPATH'] = site_dir + os.pathsep + site_dir_noarch
+        write_type_hints(site_dir)
```
this is causing a circular import dependency. Can you move it to after line 170?

BvB93 replied:
Sure thing, I'll fix during the next set of pushes.

Commit:
ENH: Partially rewrote the `generate_type_hints` script

`ctypes` scalars are now used for inferring the size of their `np.number`-based counterpart

BvB93 commented:
Xref https://github.com/numpy/numpy/issues/16548 as this issue describes more or less what this PR is doing.

BvB93 commented:
Say, aren't the numerical scalars in `ctypes` always of the same size as their `np.number`-based counterpart?
If so, then the problem with importing from `np.core` (https://github.com/numpy/numpy/pull/17514#issuecomment-713610418) can be circumvented by grabbing the scalar sizes from ctypes instead of numpy.

Any thoughts?

Commit:
REV: Removed the (now-redundant) changes to `runtest`

Commit:
TST: Updated the typing tests

Commit:
MAINT: Don't use double assignments

mypy has issues recognizing an object as valid type aliases when using the likes of `clongfloat = longcomplex = clongdouble = Any`

Commit:
MAINT: Rename `LongFloat` to `LongDouble`

Should be more consist with the `half, single, double, ...` series

Commit:
MAINT: Use `np.intp` as annotation where appropiate

Commit:
MAINT: Don't use double assignments

Commit:
MAINT: Moved the `write_type_hints()` call further up

Commit:
MAINT: Fix the `singlecomplex` import

Commit:
MAINT: Use the new charachter-code aliases in `np.dtype`

Commit:
MAINT: FIxed a rebase artifacts

Commit:
MAINT: Fixed an issue with the `complex128` key

eric-wieser commented on numpy/typing/_platform_aliases.py:
```diff
@@ -0,0 +1,62 @@
+
+Notes
+-----
+A dynamically generated stub file (with actual values) is created
+upon running setup.py.
```
If this file is replaced by the generated one at installation, what is the point in this version?

BvB93 replied:
Its purpose is to provide runtime-accessible values so we can just export them to `numpy/typing/__init__.pyi`.
This way we don't have to worry about whether `typing.TYPE_CHECKING` is `True` or `False`.

BvB93 replied:
I'd also like to note that the script run during the numpy installation generates 
`_platform_aliases.pyi`, not `_platform_aliases.py`.

eric-wieser replied:
It would be great if you could draw attention to the fact that the generated file has a different name in a comment.

eric-wieser commented on tools/generate_type_hints.py:
```diff
@@ -0,0 +1,231 @@
+
+ALLOWED_SIZES = {8, 16, 32, 64, 80, 96, 128, 256}
+
+#: A mapping with `number` aliases as keys and the name of their
+#: character-code-containing `Literal` as values.
```
If you're going to go down the route of using ctypes, this needs a clear comment about why you're doing it.

BvB93 replied:
The various comments and docstrings have been updated as of https://github.com/numpy/numpy/pull/17514/commits/9cfcea012ec17eb4c4e27480901c1ef82a8a2e20; it should now be clearer what exactly is going on with the `ctypes` approach.

eric-wieser commented on numpy/typing/_callable.py:
```diff
@@ -170,7 +172,7 @@ class _UnsignedIntBitOp(Protocol[_NBit_co]):
         @overload
         def __call__(self, __other: bool) -> unsignedinteger[_NBit_co]: ...
         @overload
-        def __call__(self, __other: int) -> signedinteger[Any]: ...
```
`type(np.uint32(1) & 1)` is `int64`, but this indicates it would be `int32`.

BvB93 replied:
Good call, fixed again as of https://github.com/numpy/numpy/pull/17514/commits/0dbc4aa45f8b5122094c3a929320ed79a065b4e6 .

Commit:
MAINT: The return type of `uint & int` is too dependant on the type of `uint`; set the return type `signedinteger[Any]`

Commit:
MAINT: Do not assume that the return precision of `x + int` is 64Bit

Use the return precision of either `self` or `int_`, depending on which one is larger

Commit:
MAINT: Provide further clarification as to what is happening in the script

* Clarified docstrings
* Clarified comments
* Use a `NamedTuple` as values in the `TO_BE_ANNOTATED` dictionary

BvB93 commented:
Rebased to get rid of a merge conflict.

hoefling commented:
Just out of curiosity - wouldn't it be easier to implement a small `mypy` plugin that transforms some types depending on the current platform?

BvB93 commented:
> Just out of curiosity - wouldn't it be easier to implement a small mypy plugin that transforms some types depending on the current platform?

In principle we could do this with the help of [`get_type_analyze_hook`](https://mypy.readthedocs.io/en/stable/extending_mypy.html#current-list-of-plugin-hooks) if we structure the types/type-aliases as following.
This does mean the users will have to enable the plugin in their mypy config files though.

``` python
from typing import Any, Annotated, Union, Literal
import numpy as np
import numpy.typing as npt

_Int64CodesBase = Literal['i8', ...]
_LongLongCodes = Literal['q', ...]
_IntPCodes = Literal['p', ...]

# A plugin can append the union with charcodes from `np.longlong`, `np.intp`, etc
_Int64Codes = Annotated[Union[_Int64CodesBase], "_Int64Codes"]

# A plugin can transform this into `intp = signedinteger[npt._64Bit]`
class intp(np.singedinteger[Any]): ...  
```

Personally I'd okay with either option, though going the plugin route does mean we'll have expose it to public API. 
I suspect the latter will be inevitable however, at last at some point in the future.

BvB93 commented:
After a bit of trial and error I managed to get the plugin-based approach working.
The respective PR is up at https://github.com/numpy/numpy/pull/17843.

BvB93 commented:
Closed in favor of https://github.com/numpy/numpy/pull/17843.
